src/ec/eddsa.rs

```diff
@@ -133,33 +174,144 @@ impl signature::VerificationAlgorithm for EdDSAParameters {
+        constant_time::verify_slices_are_equal(&r_check, &r_copy)
     }
 }
 
 impl private::Private for EdDSAParameters {}
 
 
+fn public_from_private(seed: &Seed, out: &mut PublicKey) {
+    let seed_sha512 = digest::digest(&digest::SHA512, seed);
+    let a_bytes = slice_as_scalar_ref(&seed_sha512.as_ref()).unwrap();
+    let mut a_bytes = *a_bytes;
+    unsafe {
+        GFp_ed25519_scalar_mask(&mut a_bytes);
+    }
+    let mut a = ExtPoint::new_at_infinity();
+    unsafe {
+        GFp_x25519_ge_scalarmult_base(&mut a, &a_bytes);
+        GFp_ge_p3_tobytes(out, &a);
+    }
+}
+
+fn slice_as_scalar_ref(slice: &[u8]) -> Result<&Scalar, error::Unspecified> {
+    slice_as_array_ref!(&slice[..SCALAR_LEN], SCALAR_LEN)
+}
+
+fn slice_as_scalar_ref_mut(slice: &mut [u8])
+                           -> Result<&mut Scalar, error::Unspecified> {
+    slice_as_array_ref_mut!(&mut slice[..SCALAR_LEN], SCALAR_LEN)
+}
+
 extern  {
-    fn GFp_ed25519_public_from_private(out: *mut u8/*[32]*/,
-                                       in_: *const u8/*[32]*/);
+    fn GFp_ed25519_scalar_mask(a: &mut Scalar);
+    fn GFp_ge_double_scalarmult_vartime(r: &mut Point, a_coeff: &Scalar,
+                                        a: &ExtPoint, b_coeff: &Scalar);
+    fn GFp_ge_p3_tobytes(s: &mut [u8; ELEM_LEN], h: &ExtPoint);
+    fn GFp_x25519_ge_frombytes_vartime(h: &mut ExtPoint, s: &Scalar) -> c::int;
+    fn GFp_x25519_ge_scalarmult_base(h: &mut ExtPoint, a: &Seed);
+    fn GFp_x25519_ge_tobytes(s: &mut Scalar, h: &Point);
+    fn GFp_x25519_sc_muladd(s: &mut Scalar, a: &Scalar, b: &Scalar, c: &Scalar);
+    fn GFp_x25519_sc_reduce(s: &mut Scalar);
+}
 
-    fn GFp_ed25519_sign(out_sig: *mut u8/*[64]*/, message: *const u8,
-                        message_len: c::size_t, private_key: *const u8/*[64]*/);
+// Keep this in sync with `ge_p3` in curve25519/internal.h.
+#[repr(C)]
+struct ExtPoint {
+    x: Elem,
+    y: Elem,
+    z: Elem,
+    t: Elem,
+}
 
-    fn GFp_ed25519_verify(message: *const u8, message_len: c::size_t,
-                          signature: *const u8/*[64]*/,
-                          public_key: *const u8/*[32]*/) -> c::int;
+impl ExtPoint {
+    fn new_at_infinity() -> Self {
+        ExtPoint {
+            x: [0; ELEM_LIMBS],
+            y: [0; ELEM_LIMBS],
+            z: [0; ELEM_LIMBS],
+            t: [0; ELEM_LIMBS],
+        }
+    }
+}
+
+// Keep this in sync with `ge_p2` in curve25519/internal.h.
+#[repr(C)]
+struct Point {
+    x: Elem,
+    y: Elem,
+    z: Elem,
 }
 
+impl Point {
+    fn new_at_infinity() -> Self {
+        Point {
+            x: [0; ELEM_LIMBS],
+            y: [0; ELEM_LIMBS],
+            z: [0; ELEM_LIMBS],
+        }
+    }
+}
+
+// Keep this in sync with `fe` in curve25519/internal.h.
+type Elem = [i32; ELEM_LIMBS];
+const ELEM_LIMBS: usize = 10;
+const ELEM_LEN: usize = 32;
+
+type PublicKey = [u8; PUBLIC_KEY_LEN];
+const PUBLIC_KEY_LEN: usize = ELEM_LEN;
+
+const PRIVATE_KEY_LEN: usize = ELEM_LEN;
```

Comment:
> First, we could get rid of the idea of “private key” (and PRIVATE_KEY_LEN) and just have “seed” and “scalar”. 

I really like how this shakes out in 9f7d94da1feaecb41fe330322ff465073de90905 and b5922777d3df7671fc42fd4da8799bc3070237da.

> But, it would actually be more efficient to store (a, z) and the public key in the key pair, [...]

I like this idea. I'll experiment with it as I respond to the other feedback.